### PR TITLE
fix(cli): detect shebang files in directory walks

### LINKED ISSUE #777: CLI directory mode skips extensionless shebang files
## Problem

`kakehashi format` and `kakehashi diagnose` behave differently for an explicitly named extensionless script versus the same script discovered through a directory argument.

Explicit files reach content-aware language detection, so a shebang or mode line can identify the language. Directory traversal filters entries by path/extension before reading them, so the same extensionless file is silently omitted and the command can report success with zero applicable files.

## Reproduction

1. Configure/install Lua support and a Lua formatting or diagnostic server.
2. Create an extensionless executable-style script named `tool`:

```lua
#!/usr/bin/env lua
local badly_formatted=1
```

3. Compare explicit-file and directory modes:

```sh
kakehashi format tool --check
kakehashi format . --check

kakehashi diagnose tool
kakehashi diagnose .
```

The explicit commands detect Lua from the shebang and process `tool`. The directory commands omit `tool` before reading it.

## Expected

A file discovered through directory traversal should use the same supported first-line language detection as an explicitly named file, so extensionless shebang/mode-line scripts are handled consistently.

### PR DESCRIPTION
**Pull request by atusy**
Closes #777.

## Summary
- let directory walks detect extensionless scripts from a bounded first-line prefix
- keep path-detectable files on the fast path and retain actual probe I/O failures for normal error reporting
- cap content discovery at 8 KiB so large newline-free extensionless files cannot cause unbounded discovery I/O or allocation
- preserve full-document first-line detection for explicitly named files, including mode markers beyond the directory-discovery cap
- cover format and diagnose directory shebang behavior, split UTF-8 boundaries, long unknown text, sparse binary input, and explicit late-marker fallback

## Contract
Directory discovery is intentionally bounded and may skip an extensionless file whose only language marker starts beyond 8 KiB. Explicit paths are a stronger signal and retain authoritative full-content detection.

## Validation
- `cargo test --lib language::heuristic::tests::test_detect_from_first_line::case_08_cpp_mode_line`
- `cargo test --features e2e --test e2e_cli_format e2e_directory_walk_formats_extensionless_shebang_file`
- `cargo test --features e2e --test e2e_cli_diagnose e2e_diagnose_directory_walk_includes_extensionless_shebang_file`
- `cargo clippy --lib -- -D warnings`

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved CLI directory walking to detect supported extensionless files using first-line content, including shebangs.
  * Improved handling of UTF-8 boundaries and preserved error reporting for unreadable entries.
  * Refined directory formatting and diagnosis to skip unsupported, unknown, and binary files while recognizing valid modelines.

* **Documentation**
  * Clarified when file support checks may inspect file contents.

* **Tests**
  * Added coverage for extensionless files, boundary conditions, and C++ modeline detection.

### CODE REVIEW

File: src/lsp/lsp_impl/cli/session.rs [hunk 1]
```diff
@@ -60,6 +60,36 @@ impl Kakehashi {
             .is_some()
     }
 
+    /// Whether a file discovered during a directory walk identifies a known
+    /// language by path or by its first line. The bounded read keeps filtering
+    /// cheap while matching explicit-file shebang and mode-line detection.
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/cli/session.rs]: Fixed in 5b7feabbc. The method contract now documents that actual first-line I/O failures remain candidates for normal operational-error reporting, while invalid UTF-8 is filtered as non-text.

File: tests/e2e_cli_format.rs [hunk 2]
```diff
@@ -309,6 +309,31 @@ fn e2e_directory_walk_respects_gitignore_but_explicit_path_wins() {
...
+        output.status.success(),
+        "directory format should succeed; stderr: {}",
+        String::from_utf8_lossy(&output.stderr)
+    );
+    assert!(
+        read(ws.path(), "tool").starts_with("#!/USR/BIN/ENV PYTHON"),
+        "directory walk should use shebang detection; content: {:?}; stderr: {}",
+        read(ws.path(), "tool"),
+        String::from_utf8_lossy(&output.stderr)
+    );
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/e2e_cli_format.rs:388]: Fixed in e103ba5a2. The formatted extensionless file is read once and reused by both the predicate and failure message.

File: src/lsp/lsp_impl/cli/session.rs [hunk 3]
```diff
@@ -60,6 +60,42 @@ impl Kakehashi {
...
+        if first_line.len() > MAX_FIRST_LINE_BYTES {
+            // Preserve exact explicit-file semantics without making directory
+            // filtering allocate an unbounded line: let normal processing read
+            // the full document and perform its existing content detection.
+            return true;
+        }
+        let first_line = String::from_utf8_lossy(&first_line);
+        self.language
+            .loadable_language_for_document(&path.to_string_lossy(), &first_line)
+            .is_some()
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/cli/session.rs]: Refuted on current HEAD: the lossy bounded probe no longer exists. efcc3ffc0 reads the complete first line and explicitly filters invalid UTF-8 as a non-text directory candidate. This preserves directory-walk behavior for extensionless binaries; explicit naming remains a stronger signal and may report its decoding error.

File: src/lsp/lsp_impl/cli/session.rs [hunk 4]
```diff
@@ -60,6 +60,44 @@ impl Kakehashi {
...
+        // rules include mode markers that may occur near its end, and applying
+        // a different cutoff here would recreate explicit-vs-directory drift.
+        // This runs only for path-detection misses, one file at a time, and
+        // retains less data than normal processing (which reads the full file).
+        if std::io::BufReader::new(file)
+            .read_until(b'\n', &mut first_line)
+            .is_err()
+        {
+            return true;
+        }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/cli/session.rs]: Deliberately retained and documented at current HEAD e103ba5a2. A cutoff recreates #777 for supported syntect mode markers near the line end; only path-detection misses take this path, one file at a time, and normal processing would retain the entire document. Invalid UTF-8 binaries and long unknown text are rejected after first-line detection. Exact explicit/directory correctness takes precedence over a different bounded heuristic contract.

File: src/lsp/lsp_impl/cli/session.rs [hunk 5]
```diff
@@ -60,6 +60,46 @@ impl Kakehashi {
...
+        let Ok(first_line) = std::str::from_utf8(&first_line) else {
+            // An unknown extensionless binary is not an operational read
+            // failure and cannot participate in text language detection.
+            return false;
+        };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/cli/session.rs]: Deliberately retained and documented. The suggested cap + unconditional keep was implemented earlier, but review proved it admits unknown/binary path misses and changes summaries/errors. Exact #777 parity requires the same complete first line accepted by syntect, including end-of-line mode markers. This occurs only for path misses, sequentially, and retains less than normal full-file processing.

File: src/cli/format.rs [hunk 6]
```diff
@@ -212,7 +212,7 @@ async fn run_paths(server: &Kakehashi, cwd: &Path, options: &FormatOptions) -> u
     }
 
     let collected = match collect_files(cwd, &options.paths, &options.excludes, &|path| {
-        server.cli_can_handle_path(path)
+        server.cli_can_handle_discovered_file(path)
     }) {
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/cli/format.rs:257]: Fixed with sibling documentation sweep in abbebcd9b. collect_files now states that the support callback may inspect content when the path is insufficient.

File: src/cli/diagnose.rs [hunk 7]
```diff
@@ -227,7 +227,7 @@ async fn run_paths(server: &Kakehashi, cwd: &Path, options: &DiagnoseOptions) ->
     }
 
     let collected = match collect_files(cwd, &options.paths, &options.excludes, &|path| {
-        server.cli_can_handle_path(path)
+        server.cli_can_handle_discovered_file(path)
     }) {
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/cli/diagnose.rs:244]: Fixed with sibling documentation sweep in abbebcd9b. Both collection and walk docs now describe content-aware support checks.

File: tests/e2e_cli_format.rs [hunk 8]
```diff
@@ -309,6 +309,42 @@ fn e2e_directory_walk_respects_gitignore_but_explicit_path_wins() {
     );
 }
 
+#[test]
+fn e2e_directory_walk_formats_extensionless_shebang_file() {
+    let prefix = "#!/usr/bin/env python ";
+    let padding = " ".repeat(8191 - prefix.len());
+    // Byte 8192 cuts through this multibyte character. Directory discovery
+    // must fall back to normal full-document detection instead of rejecting
+    // or silently skipping the valid UTF-8 file.
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/e2e_cli_format.rs]: Fixed in HEAD. The comment now describes BufReader internal 8 KiB chunking splitting the multibyte character while complete-line assembly preserves valid UTF-8 and shebang detection.

**Comment by atusy** [comment]
@codex review

**Comment by atusy** [comment]
@copilot review

**Comment by atusy** [comment]
@copilot review

**Comment by atusy** [comment]
@copilot review

**Review by atusy — COMMENTED** [review]
Codex exact-head review (5322249e1): no actionable findings. The bounded content probe preserves path-based discovery, detects supported first-line markers, handles truncated UTF-8 safely, and retains I/O failures for normal processing.

**Comment by atusy** [comment]
@copilot review

File: src/lsp/lsp_impl/cli/session.rs [hunk 9]
```diff
@@ -60,6 +60,57 @@ impl Kakehashi {
             .is_some()
     }
 
+    /// Whether a file discovered during a directory walk identifies a known
+    /// language by path or by a bounded first-line probe. Only path misses read
+    /// content. Files whose probe is truncated are retained for authoritative
+    /// detection during normal processing, preserving explicit-file parity
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/cli/session.rs]: Fixed at current HEAD b960db0ad: the contract now explicitly says content is read only when path-based detection fails.

File: tests/e2e_cli_format.rs [hunk 10]
```diff
@@ -309,6 +309,55 @@ fn e2e_directory_walk_respects_gitignore_but_explicit_path_wins() {
     );
 }
 
+#[test]
+fn e2e_directory_walk_formats_extensionless_shebang_file() {
+    let prefix = "#!/usr/bin/env python ";
+    let padding = " ".repeat(8191 - prefix.len());
+    // BufReader's 8 KiB internal read boundary cuts through this multibyte
+    // character. Discovery must continue assembling the complete valid UTF-8
+    // first line and still detect its shebang language.
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/e2e_cli_format.rs]: Fixed at current HEAD: the test comment now describes the hard 8 KiB boundary, split code point, and bounded-prefix behavior.

File: src/lsp/lsp_impl/cli/session.rs [hunk 11]
```diff
@@ -60,6 +60,72 @@ impl Kakehashi {
...
+    /// without retaining an arbitrarily long line. Files whose
+    /// first-line I/O fails are retained so the command's normal read path can
+    /// report the operational error; invalid UTF-8 is filtered as non-text.
+    pub(crate) fn cli_can_handle_discovered_file(&self, path: &Path) -> bool {
+        use std::io::Read as _;
+
+        if self.cli_can_handle_path(path) {
+            return true;
+        }
+        let Ok(file) = std::fs::File::open(path) else {
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/cli/session.rs:88]: Fixed in 9948e40ef by naming the lossy representation path_text.

**Comment by atusy** [comment]
@copilot review

**Review by atusy — COMMENTED** [review]
Codex exact-head review (b960db0ad): no actionable findings. The bounded content probe preserves path-based detection, adds first-line detection for extensionless discovered files, handles truncated UTF-8 safely, and retains files on I/O errors for normal error reporting.

**Comment by atusy** [comment]
@copilot review

File: src/lsp/lsp_impl/cli/session.rs [hunk 12]
```diff
@@ -60,6 +60,50 @@ impl Kakehashi {
...
+    /// from scanning arbitrarily large extensionless files; explicitly named
+    /// files remain authoritative for mode markers beyond the cap. Files whose
+    /// first-line I/O fails are retained so the command's normal read path can
+    /// report the operational error; invalid UTF-8 is filtered as non-text.
+    pub(crate) fn cli_can_handle_discovered_file(&self, path: &Path) -> bool {
+        use std::io::Read as _;
+
+        if self.cli_can_handle_path(path) {
+            return true;
+        }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/cli/session.rs:84]: Fixed in cf6d8bc80: unsupported paths with an extension now return false without opening the file.

File: src/lsp/lsp_impl/cli/session.rs [hunk 13]
```diff
@@ -60,6 +60,50 @@ impl Kakehashi {
...
+        const MAX_FIRST_LINE_BYTES: usize = 8 * 1024;
+        let path = path.to_string_lossy();
+        let mut probe = Vec::with_capacity(MAX_FIRST_LINE_BYTES + 1);
+        if std::io::BufReader::new(file)
+            .take((MAX_FIRST_LINE_BYTES + 1) as u64)
+            .read_to_end(&mut probe)
+            .is_err()
+        {
+            return true;
+        }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/cli/session.rs:102]: Fixed in cf6d8bc80: the capped reader now uses read_until and stops at the first newline.

**Comment by atusy** [comment]
@copilot review

**Review by atusy — COMMENTED** [review]
Codex exact-head review (cf6d8bc80): no actionable findings. The bounded first-line probe is applied only when path-based detection fails for extensionless files, preserves explicit-path behavior, handles truncated UTF-8 safely, and retains I/O failures for normal error reporting.

**Comment by atusy** [comment]
@copilot review

**Comment by atusy** [comment]
@copilot review

**Review by atusy — COMMENTED** [review]
Codex exact-head review after latest-main rebase (90072e337): no actionable findings. Directory discovery covers extensionless first-line markers with bounded reads and normal error reporting; focused tests cover formatting, diagnostics, UTF-8 truncation, large files, and explicit behavior.

**Comment by atusy** [comment]
@copilot review

**Review by atusy — COMMENTED** [review]
Codex exact-head review after 99adb2a00 rebase (30f44fffa): no actionable findings. Bounded first-line probing preserves path detection and handles I/O failures, invalid UTF-8, truncated multibyte sequences, and explicit files consistently.

**Comment by atusy** [comment]
@copilot review

File: tests/e2e_cli_format.rs [hunk 14]
```diff
@@ -309,6 +309,73 @@ fn e2e_directory_walk_respects_gitignore_but_explicit_path_wins() {
...
+    );
+    assert_eq!(read(ws.path(), "unknown"), unknown);
+
+    let explicit = run_format(ws.path(), &["modeline"]);
+    assert!(explicit.status.success());
+    assert!(
+        String::from_utf8_lossy(&explicit.stderr).contains("1 unchanged"),
+        "an explicit path must still use full first-line detection; stderr: {}",
+        String::from_utf8_lossy(&explicit.stderr)
+    );
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/e2e_cli_format.rs:403]: Fixed in def94be34: the explicit late-marker fixture now enables a cpp host formatter and asserts the file content is changed, which requires successful full first-line detection.

**Comment by atusy** [comment]
@copilot review

**Review by atusy — COMMENTED** [review]
Codex exact-head review (def94be34): no actionable findings. Directory discovery remains bounded and explicit late-marker behavior is now proven by an observable formatting effect.

**Comment by atusy** [comment]
@copilot review

File: src/lsp/lsp_impl/cli/session.rs [hunk 15]
```diff
@@ -60,6 +60,53 @@ impl Kakehashi {
             .is_some()
     }
 
+    /// Whether a file discovered during a directory walk identifies a known
+    /// language by path or by a bounded prefix of its first line. Content is
+    /// read only when path-based detection fails. The hard cap keeps traversal
+    /// from scanning arbitrarily large extensionless files; explicitly named
+    /// files remain authoritative for mode markers beyond the cap. Files whose
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/cli/session.rs]: Fixed in a76353b84: the contract now explicitly scopes bounded content probing to extensionless paths.

**Comment by atusy** [comment]
@copilot review

**Comment by atusy** [comment]
@copilot review

**Comment by atusy** [comment]
@copilot review

File: src/lsp/lsp_impl/cli/session.rs [hunk 16]
```diff
@@ -60,6 +60,53 @@ impl Kakehashi {
...
+        const MAX_FIRST_LINE_BYTES: usize = 8 * 1024;
+        let path_text = path.to_string_lossy();
+        let mut probe = Vec::with_capacity(MAX_FIRST_LINE_BYTES + 1);
+        if std::io::BufReader::new(file)
+            .take((MAX_FIRST_LINE_BYTES + 1) as u64)
+            .read_until(b'\n', &mut probe)
+            .is_err()
+        {
+            return true;
+        }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/cli/session.rs]: Fixed in the latest commit by applying `File::take` before constructing `BufReader`, so buffering cannot prefetch beyond the hard discovery cap.

**Comment by atusy** [comment]
@copilot review

File: tests/e2e_cli_format.rs [hunk 17]
```diff
@@ -309,6 +309,76 @@ fn e2e_directory_walk_respects_gitignore_but_explicit_path_wins() {
...
+    let source = format!("{prefix}{padding}é\nvalue = 1\n");
+    let late_mode_line = format!("{}-*- C++ -*-\nvalue = 2\n", " ".repeat(12 * 1024));
+    let ws = workspace_with(&[("tool", &source), ("modeline", &late_mode_line)]);
+    let unknown = "x".repeat(9 * 1024);
+    std::fs::write(ws.path().join("unknown"), &unknown).expect("write unknown text file");
+    let sparse_path = ws.path().join("binary");
+    let sparse = std::fs::File::create(&sparse_path).expect("create extensionless sparse binary");
+    sparse
+        .set_len(64 * 1024 * 1024)
+        .expect("size sparse binary");
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/e2e_cli_format.rs]: Fixed in `8be083793`: the sparse fixture is now 16 KiB, still above the 8 KiB discovery cap.

**Comment by atusy** [comment]
@copilot review
